src/group_theory/pi_hom.lean

```diff
@@ -0,0 +1,360 @@
+import data.fintype.card
+
+/-!
+# Canonical homomorphism from a pi monoid
+
+This file defines the construction of the canoncial homomorphism from a family of monoids.
+
+Given a family of morphisms `ϕ i : N i →* M` for each `i : I` where elements in the
+image of different morphism commute, we obtain a canoncial morphism
+`pi_hom.hom : (Π (i : I), N i) →* M` that coincides with `ϕ`
+
+## Main definitions
+
+* `pi_hom.hom : (Π (i : I), N i) →* M` is the main homomorphism
+* `pi_hom_restr.hom (S : finset S) : (Π (i : I), N i) →* M` is the homomorphism restricted to the
+   set `S`, and mostly of internal interest in this file, to allow inductive proofs.
+* `subgroup_pi_hom.hom : (Π (i : I), H i) →* G` is the specialization to `H i : subgroup G` and
+   the subgroup embedding.
+
+## Main theorems
+
+* `pi_hom.hom` conicides with `ϕ i` when restricted to `N i`
+* `pi_hom.mrange`: The range of `pi_hom.hom` is `⨆ (i : I), (ϕ i).mrange`
+* `pi_hom.range`: The range of `pi_hom.hom` is `⨆ (i : I), (ϕ i).range`
+* The range of `subgroup_pi_hom.hom` is `⨆ (i : I), H i`
+* `pi_hom.injective_of_independent`: in the case of groups, `pi_hom.hom` is injective if the `ϕ`
+   are injective and the ranges of the `ϕ` are independent.
+* `pi_hom.independent_range_of_coprime_order`: If the `N i` have coprime orders, then the ranges are
+   independent.
+* `independent_of_coprime_order`: If commuting, normal subgroups `H i` have coprime orders, they are
+   independent.
+
+-/
+
+open_locale big_operators
+open_locale classical
+
+lemma coprime_prod_left
+  {I : Type*}
+  {x : ℕ} {s : I → ℕ} {t : finset I} :
+  (∀ (i : I), i ∈ t → nat.coprime (s i) x) → nat.coprime (∏ (i : I) in t, s i) x :=
+finset.prod_induction s (λ y, y.coprime x) (λ a b, nat.coprime.mul) (by simp)
+
+-- I think it's worth keeping it and moving to appropriate file
+lemma mul_eq_one_iff_disjoint {G : Type*} [group G] {H₁ H₂ : subgroup G} :
+  disjoint H₁ H₂ ↔ ∀ {x y : G}, x ∈ H₁ → y ∈ H₂ → x * y = 1 → x = 1 ∧ y = 1 :=
+begin
+  split,
+  { intros hdis x y hx hy heq,
+    obtain rfl : y = x⁻¹ := symm (inv_eq_iff_mul_eq_one.mpr heq),
+    have hy := H₂.inv_mem_iff.mp hy,
+    have : x ∈ H₁ ⊓ H₂, by { simp, cc },
+    rw [hdis.eq_bot, subgroup.mem_bot] at this,
+    subst this,
+    simp },
+  { rintros h x ⟨hx1, hx2⟩,
+    obtain rfl : x = 1 := (h hx1 (H₂.inv_mem hx2) (mul_inv_self x)).1,
+    exact rfl, },
+end
+
+section family_of_monoids
+
+parameters {M : Type*} [monoid M]
+
+-- We have a family of monoids
+-- The fintype assumption is not always used, but declared here, to keep things in order
+parameters {I : Type*} [hfin : fintype I]
+parameters {N : I → Type*} [∀ i, monoid (N i)]
+
+-- And morphisms ϕ into G
+parameters (ϕ : Π (i : I), N i →* M)
+
+-- We assume that the elements of different morphism commute
+-- Since we need this all over the place we wrap it up in `fact`
+parameters [hcomm : fact (∀ (i j : I), i ≠ j → ∀ (x : N i) (y : N j), commute (ϕ i x) (ϕ j y))]
```

Comment:
Usually `fact` isn't for " I need this all over the place" but "I need this as an argument for something I plan to declare an `instance`.

Reply:
Should I refactor and make this an explicit argument of the various definitions? (Maybe an implicit argument for the lemmas where it can be inferred)?

Reply:
Remove the use of `fact`, and while I was it, also of `parameters`.